smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/provider/UserTuneProvider.java

```diff
@@ -0,0 +1,96 @@
+import org.jivesoftware.smack.provider.ExtensionElementProvider;
+import org.jivesoftware.smack.util.ParserUtils;
+import org.jivesoftware.smack.xml.XmlPullParser;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+import org.jivesoftware.smackx.usertune.element.UserTuneElement;
+
+/**
+ * This is the Provider Class for {@link UserTuneElement}.
+ */
+public class UserTuneProvider extends ExtensionElementProvider<UserTuneElement> {
+
+    private static final Logger LOGGER = Logger.getLogger(UserTuneProvider.class.getName());
+    public static final UserTuneProvider INSTANCE = new UserTuneProvider();
+
+    @Override
+    public UserTuneElement parse(XmlPullParser parser, int initialDepth, XmlEnvironment xmlEnvironment)
+            throws XmlPullParserException, IOException, SmackParsingException {
+
+        UserTuneElement.Builder builder = UserTuneElement.getBuilder();
+        XmlPullParser.TagEvent tag = parser.nextTag();
+        outerloop: while (true) {
+            switch (tag) {
+            case START_ELEMENT:
+                String name = parser.getName();
+                String namespace = parser.getNamespace();
+                if (!UserTuneElement.NAMESPACE.equals(namespace)) {
+                    LOGGER.log(Level.FINE, "Foreign namespace " + namespace + " detected.");
+                    continue outerloop;
+                }
+            while (tag == XmlPullParser.TagEvent.START_ELEMENT) {
+                        switch (name) {
+                        case "artist":
+                            builder.setArtist(parser.nextText());
+                            break;
+                        case "length":
+                            builder.setLength(ParserUtils.getIntegerFromNextText(parser));
+                            break;
+                        case "rating":
+                            builder.setRating(ParserUtils.getIntegerFromNextText(parser));
+                            break;
+                        case "source":
+                            builder.setSource(parser.nextText());
+                            break;
+                        case "title":
+                            builder.setTitle(parser.nextText());
+                            break;
+                        case "track":
+                            builder.setTrack(parser.nextText());
+                            break;
+                        case "uri":
+                            URI uri = ParserUtils.getUriFromNextText(parser);
+                            builder.setUri(uri);
+                            break;
+                        }
+                        tag = parser.nextTag();
+                        name = parser.getName();
+                    }
+                break;
+            case END_ELEMENT:
+                if (parser.getDepth() == initialDepth) {
+                    break outerloop;
+                }
+                break;
+            default:
```

Comment:
The `default` case can now be removed since you use `nextTag()` and all values of the returned enum are already covered in the switch-case.

Reply:
Yes Right !